ouroboros-consensus/docs/VersioningSchemeDecision.md

```diff
@@ -0,0 +1,397 @@
+- The repository as of the merge commit of a PR must not declare release versions for packages that the PR changed.
+- A package must have a release version as of a commit announced as a release of that package.
+Thus the first PR that alters a package after the previous release must change that package's declared version from the release version to a dev version.
+
+The remaining question is what should that dev version be for a specific PR?
+Suppose the previous release was `2.2.2`.
+Also suppose this PR is just a bugfix.
+A promising idea is to use `2.2.2.0` as the new dev version.
+- It has one more component than the release version, so it's distinct.
+- Standard tools typically support a variable number of components.
+- Moreover, `2.2.2.0` is greater than `2.2.2`, which seems intuitive: it's more correct than `2.2.2`.
+- Similarily, it's less than all the possible next releases: `2.2.3`, `2.3.0`, and `3.0.0`.
+
+Suppose instead the PR added a feature instead of fixing a bug.
+Is it still the case that new version should be `2.2.2.0`?
+It does seem surprising that a version `2.2.2.0` would have more _features_ than version `2.2.2`, since they both start with `2.2`.
+One could dismiss this, arguing that `2.2.2.0` is obviously a dev version (it has the extra component), and so people who understand the distinction between dev versions and release versions would know not to infer anything more than "2.2.2.0 is at least one PR ahead of version 2.2.2".
+
+However, standard tools do not understand the distinction!
+In particular, `Cabal` will happily pick `2.2.2.0` when the downstream user has written the constraint as `>= 2.2.2 && < 2.3.0`.
+In this case, because the commit that defines `2.2.2.0` adds a feature, `Cabal`'s naive inference based on `2.2.2.0 < 2.3.0` is unsound.
+One could also dismiss this concern about tooling, though, because only devs (here or downstream) should ever be building dev versions, and they should only do so explicitly (eg building from source by building a git worktree or by listing a specific git commit as the dependency).
+As long as no one installs a dev version package into a repository, there's no risk of a tool that picks amongst versions from some repository even considering a dev version.
+Dev versions are never _released_, and so should never be in any public package repository such as Hackage or [CHaP](https://github.com/input-output-hk/cardano-haskell-packages).
+The most likely way we can imagine a dev version might end up in a repository would be if a developer/user who doesn't know any better builds from source and installs a dev version into their machine's local package database.
+That single risk seems manageable.
+
+This suggests the following proposal.
+- Each release version is `A.B.C`.
+- Each dev version is `A.B.C.D`.
+- Any PR that alters a package in a way that should affect the next release version should update its version as follows.
```

Comment:
TODO. This line still incurs the concern that @michaelpj and @zliu41 raised [here](https://github.com/input-output-hk/ouroboros-network/pull/4207/#discussion_r1040345279) and https://github.com/input-output-hk/ouroboros-network/pull/4207/#discussion_r1039426571.